text/0023-starlark_load_dataset.md

```diff
@@ -0,0 +1,249 @@
+
+This variable-style import makes sense for datasets, who's names are often lengthy. Unlike package imports, we anticipate it will be very common for dataset names to collide, being distinguished only be peername. this required assign-to-name pattern will help here as well.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### differences with `load`
+As currently proposed users are expected to assign the result of `load_dataset`
+```python
+# load the `http` component of the http "package object"
+load("http.star", "http")
+# load a dataset into a variable named "fhv"
+fhv = load_dataset("b5/nyc_for_hire_vehicles")
+```
+
+In the `load` statement, the value `http` is defined by the module, which can only be known by reading the source. I find starlark's `load` statement opaque, and don't consider it a pattern we should follow.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### load into a user-supplied string
+One alternative would be to have `load_dataset` work a little closer to the way `load` does in starlark, having a user supply a second string argument that names a global variable:
+```python
+# load b5/nyc_for_hire_vehicles into a global variable named "fhv"
+load_dataset("b5/nyc_for_hire_vehicles", "fhv")
+
+# use fhv:
+fhv.get_body()
+```
+
+I attempted to code this up, it was both difficult to execute, and felt too "magical" in practice. My initial reason for specing this syntax were due to the future need to make alterations that will make the return value behave in different ways during execution. consider the following scenario:
+```python
+# only allow access to "residents" during the "transform" step. "fhv" will 
+# be equal to None during all other steps
+load_dataset("b5/nyc_for_hire_vehicles", "fhv", steps=["transform"])
+
+def download(ctx):
+  fhv.get_body() # error: cannot access 'fhv' during download step
+
+def transform(ds,ctx):
+  fhv.get_body() # this will work
+```
+
+I now realize the problem is not in how the variable is declared, it is that the value stored at `fhv` is _stateful_. How the dataset is named is an orthagonal problem. The return-value style wins out in my mind on account of reading more clearly.
+
+
+### alternate names
+We can also consider alternative names for `load_dataset`. To me the requirements are as follows:
+* show a dependency is being created
+* differ from `load`, which refers to requiring code
+* prefer concise names
+
+
+# Prior art
+[prior-art]: #prior-art
+
+We're taking a lot of inspiration for the security model for imports & dependencies from the [deno project](https://github.com/denoland/deno)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+### Loading from history
+One aspect not yet covered by this RFC is how to load previous versions of the dataset a script is affecting. The closest I have so far is supplying some sort of relative ref with no name:
+
+```python
+prev = load_dataset("~1")
```

Comment:
I don't think we should do this. At least, not without adding a solution for permissions / privilege escalation. It's not safe to assume that "~1" is a dataset meant for public consumption.